src/secure/enclave.rs

```diff
@@ -0,0 +1,156 @@
+use crate::blockcfg::{BlockBuilder, BlockDate};
+use chain_impl_mockchain::leadership::{Leader, LeaderOutput, Leadership};
+use chain_impl_mockchain::block::Block;
+use std::collections::BTreeMap;
+use std::sync::{Arc, RwLock};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct LeaderId(u32);
+
+impl LeaderId {
+    pub fn next(self) -> Self {
+        Self(self.0 + 1)
+    }
+}
+
+#[derive(Clone)]
+pub struct Enclave {
+    leaders: Arc<RwLock<BTreeMap<LeaderId, Leader>>>,
+}
+
+pub struct LeaderEvent {
+    pub id: LeaderId,
+    pub date: BlockDate,
+    pub output: LeaderOutput,
+}
+
+fn get_maximum_id<A>(leaders: &BTreeMap<LeaderId, A>) -> LeaderId {
+    leaders.keys().last().copied().unwrap_or(LeaderId(0))
+}
+
+impl Enclave {
+    pub fn new() -> Self {
+        Enclave {
+            leaders: Arc::new(RwLock::new(BTreeMap::new())),
+        }
+    }
+
+    pub fn from_vec(leaders: Vec<Leader>) -> Self {
+        let mut e = Self::new();
+        for leader in leaders {
+            e.add_leader(leader);
+        }
+        e
+    }
+
+    pub fn get_leaderids(&self) -> Vec<LeaderId> {
```

Comment:
Nitpick: name `get_leader_ids` for snake case consistency.

Reply:
this will go away once this is used as an abstract box